### PR TITLE
Comments

### PR DESCRIPTION
**Pull request by alejoaquili**
## Summary
A FlatList of Comments for book details screen implemented with conditional rendering.

#### Nexus S
Peek 2019-03-28 08-25

#### Nexus 4
Peek 2019-03-28 08-27

#### Nexus 5X
Peek 2019-03-28 08-22

#### Nexus 6
Peek 2019-03-28 08-30

## Trello Card
https://trello.com/c/nM6XCeLP/20-n-72-tabla-de-comentarios

### CODE REVIEW

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import Comment from './components/Comment';
+import styles from './styles';
+
+class CommentList extends Component {
+
+    constructor(props) {
+        super(props);
+        this.state = { isViewedAll: false };
+    }
```
Comment (guidoprinc) [inline src/app/screens/BookDetail/components/CommentList/index.js]: we try to never use constructor? just add a line like:
`state = { isViewedAll: false }`

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    renderItem = ({item}) =>  <Comment username={item.user.name}  userImageSource={item.user.imageSource} comment={item.comment}/>;
```
Comment (guidoprinc) [inline src/app/screens/BookDetail/components/CommentList/index.js]: remember the spaces: `({ item })`

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    keyExtractor = (item) => `${item.id}`;
```
Comment (guidoprinc) [inline src/app/screens/BookDetail/components/CommentList/index.js]: if it's just an element, you don't need the parenthesis

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    onPressHandler = () => (this.setState(previousState => ({ isViewedAll: true })));
+    render() {
```
Comment (guidoprinc) [inline src/app/screens/BookDetail/components/CommentList/index.js]: space between function and render. Also remove parenthesis for the setState

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        const {comments} = this.props;
```
Comment (guidoprinc) [inline src/app/screens/BookDetail/components/CommentList/index.js]: spaces

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        return (
+            <FlatList
+            data={ (this.state.isViewedAll) ? comments.filter((x,i) => [0,1,2,3,4].includes(i)) : comments.filter((x,i) => [0,1].includes(i))}
+            renderItem={this.renderItem}
+            keyExtractor={this.keyExtractor}
+            contentContainerStyle={styles.container}
+            ListFooterComponent={() =><View>{(this.state.isViewedAll) ? <Text></Text> : <TouchableOpacity style={styles.footer} onPress={this.onPressHandler} >{(comments.length <= 2) ? <Text></Text> : <Text style={styles.footerText}>View All</Text>}</TouchableOpacity>}</View>}
```
Comment (guidoprinc) [inline src/app/screens/BookDetail/components/CommentList/index.js]: why do you have an empty Text?

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    renderItem = ({item}) =>  <Comment username={item.user.name}  userImageSource={item.user.imageSource} comment={item.comment}/>;
+
+    keyExtractor = (item) => `${item.id}`;
+
+    onPressHandler = () => (this.setState(previousState => ({ isViewedAll: true })));
+    render() {
+        const {comments} = this.props;
+        return (
+            <FlatList
+            data={ (this.state.isViewedAll) ? comments.filter((x,i) => [0,1,2,3,4].includes(i)) : comments.filter((x,i) => [0,1].includes(i))}
```
Comment (guidoprinc) [inline src/app/screens/BookDetail/components/CommentList/index.js]: spaces

File: src/app/screens/BookDetail/index.js [hunk 8]
```diff
@@ -3,13 +3,18 @@ import { View } from 'react-native';
...
 class BookDetail extends Component {
     render() {
       const book = this.props.navigation.getParam('book');
+      // TODO: remove test.
+
       return (
         <View style={styles.container}>
           <BookDetailComponent title= {book.title} author={book.author} year={book.year} type={book.type} imageSource={book.imageSource}/>
+          <CommentList comments={book.comments}></CommentList>
+
```
Comment (guidoprinc) [inline src/app/screens/BookDetail/index.js]: if you only have the render, maybe you can use a stateless function

File: src/app/screens/BookDetail/components/BookDetail/styles.js [hunk 9]
```diff
@@ -62,11 +62,13 @@ export default StyleSheet.create({
     },
     addToWishListText: {
         fontSize: SMALL,
-        color: skyBlue
+        color: skyBlue,
+        fontWeight: "bold"
```
Comment (sfernandez11) [inline src/app/screens/BookDetail/components/BookDetail/styles.js]: Extract the "bold" constant to `@constants/fontSize`

File: src/app/screens/BookDetail/components/CommentList/components/Comment/index.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import React, { Component } from 'react';
+import { View, Image, Text } from 'react-native';
+
+import styles from './styles';
+
+function Comment({username, userImageSource, comment}) {
+    return (
+        <View style={styles.container}>
+            <Image style={styles.image} source={userImageSource}/>
```
Comment (sfernandez11) [inline src/app/screens/BookDetail/components/CommentList/components/Comment/index.js]: Remember de `resizeMode`

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+    constructor(props) {
+        super(props);
+        this.state = { isViewedAll: false };
+    }
+    renderItem = ({item}) =>  <Comment username={item.user.name}  userImageSource={item.user.imageSource} comment={item.comment}/>;
+
+    keyExtractor = (item) => `${item.id}`;
+
+    onPressHandler = () => (this.setState(previousState => ({ isViewedAll: true })));
```
Comment (sfernandez11) [inline src/app/screens/BookDetail/components/CommentList/index.js]: `handlePressViewAll = () => this.setState({ isViewedAll: true })`

File: src/app/screens/BookDetail/components/CommentList/index.js (hunk 7 above)
Comment (sfernandez11) [reply src/app/screens/BookDetail/components/CommentList/index.js]: Destruct the `this.state` object as you do with `this.props`:
```
const { isViewedAll } = this.state;
data={ isViewedAll ? comments.filter((x,i) => [0,1,2,3,4].includes(i)) : comments.filter((x,i) => [0,1].includes(i))}
```
Also, your are repeating this `comments.filter((x,i) => ARRAY.includes(i))`. I think you can extract this functions and pass via params the `comments` array and the `indexs` array ([0,1,2,3,4] and [0,1])

File: src/app/screens/BookDetail/components/CommentList/index.js (hunk 6 above)
Comment (sfernandez11) [reply src/app/screens/BookDetail/components/CommentList/index.js]: Also, you can extract this function like you do with `renderItem` and `keyExtractor`. Avoid use arrow functions inside render methods

File: src/app/screens/BookDetail/components/CommentList/index.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        const { comments } = this.props;
+        const { isViewedAll } = this.state;
+
+        return (
+            <FlatList
+            data={ (isViewedAll) ? this.getCommentsByIndex(comments, [0, 1, 2, 3, 4]) : this.getCommentsByIndex(comments, [0, 1])}
+            renderItem={this.renderItem}
+            keyExtractor={this.keyExtractor}
+            contentContainerStyle={styles.container}
+            ListFooterComponent={<View>{this.renderViewAll(comments, isViewedAll)}</View>}
```
Comment (sfernandez11) [inline src/app/screens/BookDetail/components/CommentList/index.js]: Why the View here? I think it's unnecessary

**Pull request merged**